addons/mail/static/src/js/mail.js

```diff
@@ -1489,70 +1447,64 @@ openerp.mail = function (session) {
+            var res = [];
+            _(messages).each(function(data){
+                data.thread_level = self.thread_level || 0;
+                data.options = _.extend(data.options || {}, self.options);
+                if (data.type=='expandable') {
+                    var message = new mail.ThreadExpandable(self, data, {'context':{
+                        'default_model': data.model || self.context.default_model,
+                        'default_res_id': data.res_id || self.context.default_res_id,
+                        'default_parent_id': self.id,
+                    }});
+                } else {
+                    data.record_name= (data.record_name != '' && data.record_name) || (self.parent_message && self.parent_message.record_name);
+                    var message = new mail.ThreadMessage(self, data, {'context':{
+                        'default_model': data.model,
+                        'default_res_id': data.res_id,
+                        'default_parent_id': data.id,
+                    }});
                 }
-            }
-            self.messages.push( message );
-
-            return message;
+                // check if the message is already create
+                self.messages.filter(function(msg) { return msg.id != message.id; })
+                self.messages.push(message);
+                res.push(message);
+            });
+            return res;
         },
 
         /**
-         *insert the message on the DOM.
-         * All message (and expandable message) are sorted. The method get the
-         * older and newer message to insert the message (before, after).
-         * If there are no older or newer, the message is prepend or append to
-         * the thread (parent object or on root thread for flat view).
-         * The sort is define by the thread_level (O for newer on top).
-         * @param : {object} ThreadMessage object
+         * render thread: append all messages to thead into DOM
          */
-        insert_message: function (message, dom_insert_after, prepend) {
-            var self=this;
-            if (this.options.show_compact_message > this.thread_level) {
-                this.instantiate_compose_message();
-                this.compose_message.do_show_compact();
-            }
-
-            this.$('.oe_view_nocontent').remove();
+        render_thread: function(messages,dom_insert_after, prepend){
+            var self = this;
             if (dom_insert_after && dom_insert_after.parent()[0] == self.$el[0]) {
-                message.insertAfter(dom_insert_after);
+                _(messages).invoke('insertAfter', dom_insert_after)
             } else if (prepend) {
-                message.prependTo(self.$el);
+                _(messages).invoke('prependTo', self.$el)
             } else {
-                message.appendTo(self.$el);
+                _(messages).invoke('appendTo', self.$el)
             }
-            message.$el.hide().fadeIn(500);
```

Comment:
yes , why we will wait 500 millisecond to display thread, staring as soon as messages arrived